File: Python/sources/Foundations/Algebra/Fourier.py
import matplotlib.pyplot as plt
from math  import  pi, sin  
from cmath  import exp  
from numpy import  array, conj  



def Fourier_example(): 
  
     print( " Fourier series examples    " ) 

     N = 24     # Fourier truncated series 
     M = 2000   # points to plot 

     x = array( [ 2*pi*i/M for i in range(M) ] ) 
     y = array( list( map( lambda x: Fourier_series( N, x ) , x) ) ) 

     plt.plot(x, y) 
     plt.show() 

     y = array( list( map( lambda x: sine_Fourier( N, x ) , x) ) ) 

     plt.plot(x, y) 
     plt.show() 

#--------------------------------------------------
# f(x) = sum from k= -N/2 to N/2 c_k exp( I k x ) 
#      
#         with  c_k = ( a_k - I * b_k ) / 2, k>0 
#               c_{-k} = conjugate( c_k ) 
#
#               a_k : cosine coefficients 
#               b_k : sine coefficients 
#--------------------------------------------------
def Fourier_series(N, x): 
   
     
    M = int(N/2)
   
#   complex :: c(-N/2:N/2), e(-N/2:N/2) 
#   real :: a(0:N/2), b(0:N/2) 
    a = array( [ 0. for k in range(M+1) ] )
    b = array( [ 0. for k in range(M+1) ] ) 
    # WARNING DOES NOT WORK  if array is initialized with 0 instead of 0. 
    c = array( [ complex(0., 0.)   for k in range(-M, M+1) ] )
    e = array( [ exp( 1j * k * x ) for k in range(-M, M+1) ] )

    for k in range(1, M+1): 
            b[k] = 2*(-1)**(k+1) / (pi*k) 

#   c(0:N/2)   =   ( ak - I * bk )/2 
#   c(-N/2:-1) = conjg( c(N/2:1:-1) )
#   WARNING in python c[0:3]= c[0], c[1], c[2] 
#   Python is a pain in the ass with indexes starting at 0. 
    for k in range(0, M+1): 
        c[M+k] = ( a[k] - b[k]* 1j )/2 
        
    for k in range(0, M+1): 
        c[k] = conj( c[N-k] )
   
    f = sum( c * e ).real 

    return f
    

def sine_Fourier(N, x): 


    M = int(N/2)

  # k goes from 0 to M-1
  # It obligues to modify k by k+1
    b = array( [ 2*(-1)**(k+2) / (pi*(k+1)) for k in range(M) ] ) 
    
    sink = array( [ sin( (k+1) * x ) for k in range(M) ] ) 
    
    f = sum( b * sink )  
    
    return f 

File: Fortran/sources/Advanced_programming/functional programming/Fourier.f90
module Fourier 
    
    use dislin 
    implicit none 
    
    real, parameter :: PI = 4 * atan(1d0) 
    complex, parameter :: I = (0, 1)
    
    contains
    
subroutine Fourier_example 
    
    integer :: N = 24   ! Fourier truncated series 
    integer, parameter :: M = 2000 ! points to plot 
    real ::  x(0:M), y(0:M)
    integer :: i
   
    x = [ ( 2 * PI * i /M, i=0, M) ] 
   
    call scrmod("reverse")
    y = Fourier_series( N, x ) 
    call qplot(x, y, M+1) 
    
    y = sine_Fourier( N, x ) 
    call qplot(x, y, M+1) 
    
end subroutine

!--------------------------------------------------
! f(x) = sum from k= -N/2 to N/2 c_k exp( I k x ) 
!      
!         with  c_k = ( a_k - I * b_k ) / 2, k>0 
!               c_{-k} = conjugate( c_k ) 
!
!               a_k : cosine coefficients 
!               b_k : sine coefficients 
!--------------------------------------------------
elemental function Fourier_series(N, x) result(f) 
    integer, intent(in) :: N 
    real, intent(in) :: x 
    real :: f 
   
    complex :: c(-N/2:N/2), e(-N/2:N/2) 
    real :: a(0:N/2), b(0:N/2) 
    integer :: k 
    
    a = 0 
    b(0) = 0 
    b(1:N/2) = [ (2*(-1)**(k+1) / (PI*k), k=1, N/2 ) ] 
    
    c(0:N/2)   =   ( a - I * b )/2 
    c(-N/2:-1) = conjg( c(N/2:1:-1) )
    
    e = [ ( exp( I * k * x ), k=-N/2, N/2 ) ] 
      
    f = real ( sum( c * e ) )
    
    
end function 

elemental function sine_Fourier(N, x) result(f) 
    integer, intent(in) :: N 
    real, intent(in) :: x 
    real :: f 
   
    real :: b(1:N/2), sink(1:N/2)  
    integer :: k 
    
    b = [ (2*(-1)**(k+1) / (PI*k), k=1, N/2 ) ] 
    
    sink = [ ( sin( k * x ), k=1, N/2 ) ] 
    
    f = sum( b * sink ) 
    
end function 


    
end module 
    
